### ISSUE TITLE
Potential bug in SqlAlchemy migrations

### ISSUE THREAD

**Issue by sphuber** [aiidateam/aiida-core#4590]
I have a production SqlAlchemy database that fails to migrate with `aiida-core==1.5.0`. It starts at revision `118349c10896`, migrates successfully to `bf591f31dd12`, but then fails to migrate to revision `0edcdd5a30f0` with the exception:
```
Traceback (most recent call last):
  File "/home/sph/.virtualenvs/aiida_repo/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1278, in _execute_context
    cursor, statement, parameters, context
  File "/home/sph/.virtualenvs/aiida_repo/lib/python3.7/site-packages/sqlalchemy/engine/default.py", line 593, in do_execute
    cursor.execute(statement, parameters)
psycopg2.errors.ObjectInUse: cannot ALTER TABLE "db_dbgroup" because it has pending trigger events


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/sph/.virtualenvs/aiida_repo/bin/verdi", line 12, in <module>
    sys.exit(verdi())
  File "/home/sph/.virtualenvs/aiida_repo/lib/python3.7/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/home/sph/.virtualenvs/aiida_repo/lib/python3.7/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/home/sph/.virtualenvs/aiida_repo/lib/python3.7/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/sph/.virtualenvs/aiida_repo/lib/python3.7/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/sph/.virtualenvs/aiida_repo/lib/python3.7/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/sph/.virtualenvs/aiida_repo/lib/python3.7/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/home/sph/code/aiida/env/repo/aiida-core/aiida/cmdline/commands/cmd_database.py", line 43, in database_migrate
    backend.migrate()
  File "/home/sph/code/aiida/env/repo/aiida-core/aiida/orm/implementation/sqlalchemy/backend.py", line 47, in migrate
    self._schema_manager.migrate()
  File "/home/sph/code/aiida/env/repo/aiida-core/aiida/backends/manager.py", line 149, in migrate
    self._migrate_database_version()
  File "/home/sph/code/aiida/env/repo/aiida-core/aiida/backends/sqlalchemy/manager.py", line 151, in _migrate_database_version
    command.upgrade(config, 'head')
  File "/home/sph/.virtualenvs/aiida_repo/lib/python3.7/site-packages/alembic/command.py", line 298, in upgrade
    script.run_env()
  File "/home/sph/.virtualenvs/aiida_repo/lib/python3.7/site-packages/alembic/script/base.py", line 489, in run_env
    util.load_python_file(self.dir, "env.py")
  File "/home/sph/.virtualenvs/aiida_repo/lib/python3.7/site-packages/alembic/util/pyfiles.py", line 98, in load_python_file
    module = load_module_py(module_id, path)
  File "/home/sph/.virtualenvs/aiida_repo/lib/python3.7/site-packages/alembic/util/compat.py", line 184, in load_module_py
    spec.loader.exec_module(module)
  File "<frozen importlib._bootstrap_external>", line 728, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/home/sph/code/aiida/env/repo/aiida-core/aiida/backends/sqlalchemy/migrations/env.py", line 75, in <module>
    run_migrations_online()
  File "/home/sph/code/aiida/env/repo/aiida-core/aiida/backends/sqlalchemy/migrations/env.py", line 68, in run_migrations_online
    context.run_migrations()
  File "<string>", line 8, in run_migrations
  File "/home/sph/.virtualenvs/aiida_repo/lib/python3.7/site-packages/alembic/runtime/environment.py", line 846, in run_migrations
    self.get_context().run_migrations(**kw)
  File "/home/sph/.virtualenvs/aiida_repo/lib/python3.7/site-packages/alembic/runtime/migration.py", line 520, in run_migrations
    step.migration_fn(**kw)
  File "/home/sph/code/aiida/env/repo/aiida-core/aiida/backends/sqlalchemy/migrations/versions/0edcdd5a30f0_dbgroup_extras.py", line 37, in upgrade
    op.alter_column('db_dbgroup', 'extras', nullable=False)
  File "<string>", line 8, in alter_column
  File "<string>", line 3, in alter_column
  File "/home/sph/.virtualenvs/aiida_repo/lib/python3.7/site-packages/alembic/operations/ops.py", line 1779, in alter_column
    return operations.invoke(alt)
  File "/home/sph/.virtualenvs/aiida_repo/lib/python3.7/site-packages/alembic/operations/base.py", line 374, in invoke
    return fn(self, operation)
  File "/home/sph/.virtualenvs/aiida_repo/lib/python3.7/site-packages/alembic/operations/toimpl.py", line 56, in alter_column
    **operation.kw
  File "/home/sph/.virtualenvs/aiida_repo/lib/python3.7/site-packages/alembic/ddl/postgresql.py", line 147, in alter_column
    **kw
  File "/home/sph/.virtualenvs/aiida_repo/lib/python3.7/site-packages/alembic/ddl/impl.py", line 178, in alter_column
    existing_comment=existing_comment,
  File "/home/sph/.virtualenvs/aiida_repo/lib/python3.7/site-packages/alembic/ddl/impl.py", line 140, in _exec
    return conn.execute(construct, *multiparams, **params)
  File "/home/sph/.virtualenvs/aiida_repo/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1014, in execute
    return meth(self, multiparams, params)
  File "/home/sph/.virtualenvs/aiida_repo/lib/python3.7/site-packages/sqlalchemy/sql/ddl.py", line 72, in _execute_on_connection
    return connection._execute_ddl(self, multiparams, params)
  File "/home/sph/.virtualenvs/aiida_repo/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1076, in _execute_ddl
    compiled,
  File "/home/sph/.virtualenvs/aiida_repo/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1318, in _execute_context
    e, statement, parameters, cursor, context
  File "/home/sph/.virtualenvs/aiida_repo/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1512, in _handle_dbapi_exception
    sqlalchemy_exception, with_traceback=exc_info[2], from_=e
  File "/home/sph/.virtualenvs/aiida_repo/lib/python3.7/site-packages/sqlalchemy/util/compat.py", line 178, in raise_
    raise exception
  File "/home/sph/.virtualenvs/aiida_repo/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1278, in _execute_context
    cursor, statement, parameters, context
  File "/home/sph/.virtualenvs/aiida_repo/lib/python3.7/site-packages/sqlalchemy/engine/default.py", line 593, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.OperationalError: (psycopg2.errors.ObjectInUse) cannot ALTER TABLE "db_dbgroup" because it has pending trigger events

[SQL: ALTER TABLE db_dbgroup ALTER COLUMN extras SET NOT NULL]
(Background on this error at: http://sqlalche.me/e/13/e3q8)
```

Interestingly, both `bf591f31dd12` and `0edcdd5a30f0` alter the `db_dbgroup` table. The exception suggests that when the second migration comes around, the previous one has not completely finished or at least left unfinished trigger events on the table. This prompted me to try and perform the migrations one-by-one using the command line tool that ships with `aiida-core`:
```
python aiida/backends/sqlalchemy/manage.py upgrade bf591f31dd12
python aiida/backends/sqlalchemy/manage.py upgrade 0edcdd5a30f0
```
and lo and behold, it works.
So I think there is something wrong in the migration, either the way we defined them or a bug in alembic, where the first migration is not properly finished/flushed/whatever before the second one starts. No idea why though.

I tried splitting the revision `0edcdd5a30f0` into two revisions moving the line `op.alter_column('db_dbgroup', 'extras', nullable=False)` to the second revision, however, the exact same exception was thrown. It seems then that the problem has to do with all revisions happening in the same transaction. When we migrate manually one revision at a time, each revision is done in its own separate transaction. However, we set the flag `transaction_per_migration` to `True` when configuring the alembic environment, so that should enforce using a single transaction for each revision. (This actually emulates the behavior of the database migrations which is why we set it.)

**Comment by sphuber** [comment]
@giovannipizzi any ideas?

**Comment by sphuber** [comment]
@chrisjsewell and @ramirezfranciscof this is an important issue to nail down because if it is real, it could cause a lot of problems with actual users trying to migrate. Would any of you be able to investigate this as well. I have a database dump that reproduces it, but it is quite big. Let me know if you can work on it and then we can see how to share it.

**Comment by sphuber** [comment]
I have created a MWE database that reproduces the problem. Essentially it is a clean SQLA database with a single user and a single `Group` instance that is at revision `118349c10896`. The SQL dump is attached in a zip file (cannot upload `.psql` files directly). If you try to migrate this database with `aiida-core==1.5.0` you should get the problem above. This to me suggests that the database migration is broken for _all_ SqlAlchemy databases for `aiida-core==1.5.0` so I am moving this priority to critical, as this really needs to be fixed.

[aiida_broken_sqla_db.zip](https://github.com/aiidateam/aiida-core/files/5607802/aiida_broken_sqla_db.zip)

**Comment by chrisjsewell** [comment]
Thanks yeh I will try to fave a look in the next few days (after releasing quantum mobile and creating the Conda pkg for the disk/objectstore)

**Comment by giovannipizzi** [comment]
This seems interesting:
https://stackoverflow.com/questions/12838111/django-db-migrations-cannot-alter-table-because-it-has-pending-trigger-events
(and actually comes from the same error thrown by Django)
The solution seems to suggest that the error might be misleading, and the issue is that you might have some `null` fields in the column you're setting as NULL? Is this possible?

**Comment by sphuber** [comment]
It could, but it shouldn't, because before making the column non-nullable we [set all `extras` to an empty dictionary](https://github.com/aiidateam/aiida-core/blob/716a1d8f6801bd9ba72162a73497e1085a64bc52/aiida/backends/sqlalchemy/migrations/versions/0edcdd5a30f0_dbgroup_extras.py#L33). Note again that if I run the migration one step at a time, it works just fine. So each individual migration should be ok.

**Comment by giovannipizzi** [comment]
Ah ok thanks, this is very useful.
Is then the problem then simply here?

https://github.com/aiidateam/aiida-core/blob/855ae82e42b8a50e6c507fe9083187a22fe2cfea/aiida/backends/sqlalchemy/migrations/env.py#L60-L68

We're setting `transaction_per_migration=True, but then wrapping *again* in another transaction, that is committed only at the very end.

I would try to replace
```python
with context.begin_transaction(): 
     context.run_migrations() 
```
with simply 
```
context.run_migrations() 
```
and check (add test?)

1. that indeed the problem disappears
2. that still each migration is happening in a transaction, so that if something fails, partial migrations are reverted

**Comment by sphuber** [comment]
I saw the same thing when reading up on `transaction_per_migration` and tried removing the outside transaction but that didn't help unfortunately.

**Comment by giovannipizzi** [comment]
Some really wrong is indeed going on (or maybe I don't understand it). Here are my pointers but I won't have much more time today to debug this.

I replaced
```python
with context.begin_transaction(): 
     context.run_migrations() 
```
with 
```python
print("BEFORE")
context.run_migrations() 
print("AFTER")
```

Now, the problem is that run_migrations is
1. called even before running th migrations! I get:
```$ verdi -p test_sqla shell
/BEFORE
AFTER
Critical: 
Database schema version `041a79fc615f` is incompatible with the required schema version `0edcdd5a30f0`.
To migrate the database schema version to the current one, run the following command:

    verdi -p test_sqla database migrate
```

If I actually run the migration, I get:
```
$ verdi -p test_sqla database migrate -f
BEFORE
AFTER
BEFORE
AFTER
BEFORE
AFTER
BEFORE
Traceback (most recent call last):
...
```
(I wouldn't worry about the error at this stage, this is a test DB); but as you see the function is called *many* times.

This definitely comes (at least) from the fact that the code is in the "main" part of the script, not in a function, so at least at every import it's run. Then I don't know if it's expected that it's called multiple times.

Anyway, indeed this is something very strange.
I imagine this file is implicitly called by alembic. But maybe then we shouldn't keep it in an importable file in the AiiDA source code?

Also:
https://github.com/aiidateam/aiida-core/blob/02248cf3686a0ab89faf1625e0da24d9e33d8cde/aiida/backends/sqlalchemy/manager.py#L48
is it expected that `ALEMBIC_FILENAME = 'alembic.ini'` is not existing? (Maybe this is OK)

**Comment by giovannipizzi** [comment]
BTW, I think it's not needed to remove the `with context.begin_transaction():`:
As it says [here](https://alembic.sqlalchemy.org/en/latest/api/runtime.html?highlight=transaction_per_migration#alembic.runtime.migration.MigrationContext.begin_transaction)
> A “Logical” transaction means that the operation may or may not correspond to a real database transaction. If the target database supports transactional DDL (or EnvironmentContext.configure.transactional_ddl is true), the EnvironmentContext.configure.transaction_per_migration flag is not set, and the migration is against a real database connection (as opposed to using “offline” --sql mode), a real transaction will be started.

So I imagine in our case it's just a no-op and better to keep. I would investigate better why it's called multiple times

**Comment by giovannipizzi** [comment]
I looked at the stack, the one "calling" it each time is the `get_schema_version_database` call, that calls `with self.alembic_config()`. Not sure if this is the problem.

**Comment by giovannipizzi** [comment]
OK - apparently, when you call `with EnvironmentContext(config, script, fn=some_function):`
the intended behaviour is to call the env.py script, and the run_migrations will actually run `fn`... e.g. exactly to get the current revision. So probably my guess is wrong. Not sure what to test now... I'm going to stop - it would be useful probably to print all connections that went into PostgreSQL during the migration, to see what is actually being asked to the DB, and if/how many transactions are actually performed. I think @chrisjsewell knows how to do it

**Comment by sphuber** [comment]
I have been wrestling with this all night and I have some bad news and good news:

* bad news: I still don't know why the pending trigger events are happening or how to solve it
* good news: I found another way to write the migration that works and doesn't suffer from the same problem

In doing so, I have also cleaned up and simplified the code using alembic a lot and opened a PR for all of it: #4602 
If we cannot find out soon what the source of the pending trigger events is, I think we may have to go with this solution such that databases can actually be properly migrated. Not really satisfying though that we still understand all of this so badly...

**Comment by sphuber** [comment]
Fixed in PR #4602
